controllers/upgrade_controller.go

```diff
@@ -0,0 +1,231 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	mellanoxv1alpha1 "github.com/Mellanox/network-operator/api/v1alpha1"
+	"github.com/Mellanox/network-operator/pkg/consts"
+	"github.com/Mellanox/network-operator/pkg/upgrade"
+)
+
+const OfedDriverLabel = "nvidia.com/ofed-driver"
+
+// UpgradeReconciler reconciles OFED Daemon Sets for upgrade
+type UpgradeReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// ResourceType holds information about k8s resource type
+type ResourceType struct {
+	APIVersion string
+	Kind       string
+}
+
+//nolint
+// +kubebuilder:rbac:groups=mellanox.com,resources=*,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;update;patch
+// +kubebuilder:rbac:groups="",resources=pods,verbs=list
+// +kubebuilder:rbac:groups=apps,resources=deployments;daemonsets;replicasets;statefulsets,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *UpgradeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("upgrade", req.NamespacedName)
+	reqLogger.V(consts.LogLevelInfo).Info("Reconciling Upgrade")
+
+	nicClusterPolicy := &mellanoxv1alpha1.NicClusterPolicy{}
+	err := r.Get(context.TODO(), types.NamespacedName{Name: consts.NicClusterPolicyResourceName}, nicClusterPolicy)
+
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	upgradePolicy := nicClusterPolicy.Spec.OfedUpgradePolicy
+	if upgradePolicy == nil {
+		reqLogger.V(consts.LogLevelInfo).Info("OFED Upgrade Policy is not defined, skipping driver upgrade")
+		return ctrl.Result{}, nil
+	}
+
+	state, err := r.BuildState(ctx, upgradePolicy)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	reqLogger.V(consts.LogLevelDebug).Info("Built upgrade state", "state", state)
+
+	return ctrl.Result{}, nil
+}
+
+func (r *UpgradeReconciler) BuildState(
```

Comment:
Nit: can you add docstring to the methods ?
(general comment for the entire PR)